clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -368,7 +368,8 @@ CodeGenModule::CodeGenModule(ASTContext &C,
   IntTy = llvm::IntegerType::get(LLVMContext, C.getTargetInfo().getIntWidth());
   IntPtrTy = llvm::IntegerType::get(LLVMContext,
     C.getTargetInfo().getMaxPointerWidth());
-  Int8PtrTy = llvm::PointerType::get(LLVMContext, 0);
+  Int8PtrTy = llvm::PointerType::get(LLVMContext,
```

Comment:
This seems to cause regressions for HIP -fsanitize=address.

Basically rocm device library for ASAN is compiled from OpenCL source code, for which the default address space is mapped to addr space 5 in IR. Int8PtrTy  is used to determine the pointer type in the llvm.compiler.used global array, which causes the pointers in that array to be in addr space 5. However, for HIP those are in addr space 0. The variable from different bitcode are appended together by the linker. The linker checks their type and emits error since they do not match.

https://godbolt.org/z/s48fTj7Pv

Before this change, the pointers are in addr space 0 for both HIP and OpenCL, therefore no such issue.

Reply:
This seems to be an error on ASAN's side, it's relying on non-guaranteed behaviour and linking BC emanating from different languages with different AS maps. To wit, 0 is private for OCL, so having `llvm.used` and `llvm.compiler.used` be global arrays of pointers to private is somewhat suspect. I will / should fix `emitUsed` to rely on GlobalsInt8PtrTy, which seems like the right thing to do anyway, but whilst that'll make ASAN "work", it doesn't make ASAN correct, IMHO.

Reply:
The AMDGPU address space mapping is at https://llvm.org/docs/AMDGPUUsage.html#address-spaces and address space 0 is not private.  The address space for all languages compiled for the AMDGPU target is consistent.

Reply:
The IR address space is a pure target concept. Any address space error would be on the frontend emitting the wrong IR for the target

Reply:
Fixing emitUsed probably is a feasible solution, but GlobalsInt8PtrTy will cause the used array containing pointers to addr space 1, which is not backward compatible for HIP. Maybe just use pointer to addr space 0 type as before.

Reply:
llvm.used should be using the default globals addrspace. I believe there is special case linker code to handle mismatched address spaces for the intrinsic globals, so I don't think fixing this would break anything

Reply:
> Fixing emitUsed probably is a feasible solution, but GlobalsInt8PtrTy will cause the used array containing pointers to addr space 1, which is not backward compatible for HIP. Maybe just use pointer to addr space 0 type as before.

This would be wrong, assuming that 0 / Unqual is generally equivalent with some form of generic pointer is the core problem here. I'm not exactly sure what you mean by "backward compatible" here though, both the HIP source and whatever BC gets linked should go through the same toolchain, surely? Also, what @arsenm said sounds about right.

Reply:
> The IR address space is a pure target concept. Any address space error would be on the frontend emitting the wrong IR for the target

I don't think mixing languages with different LangAS maps is sound, OCL originating BC is not generic BC. There's no "error" here, the FE is doing what is asked of it (indiscriminately assuming that AS 0 is some generic/flat pointer that is valid everywhere is rather risque, as per prior conversation in this review). The actual problem is, AFAICT, that when we call `AMDGPUTargetInfo::adjust` we indiscriminately set private as default for any and all OCL compilations? There's a 6 year old `TODO`, which may or may not be vestigial at this point.

Reply:
> I don't think mixing languages with different LangAS maps is sound

It is entirely sound, and required for this entire system to work (e.g. we implement the libraries in OpenCL used by all the languages). The point of the LangAS is to map to the target address space, which does not care what the language is. Different languages should be trying to be ABI compatible with each other, which includes emitting the correct IR address space

Reply:
> > I don't think mixing languages with different LangAS maps is sound
> 
> It is entirely sound, and required for this entire system to work

I am aware of what we decided to do. Whether or not that is actually sound is a different kettle of fish, but let's agree to disagree and move on. Stepping back, I'm trying to figure out what the actual suggestion is. Reverting the change is wrong, because there are other targets where for which 0 is definitely not a sound default, and this impacts more than the `used`/`compiler.used` arrays. Seems like there are two viable options:

1. I'm going to propose changing `emitUsed` anyway, since that makes sense even if the AS is indeed mostly meaningless there
2. IMHO OCL should default to `AMDGPUDefIsGenMap` for its LangAS map, which would "fix" this as well.
 
If there's some third reasonable solution, I apologise for missing it - please share!